pkg/client/results/processing.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright the Sonobuoy contributors 2019
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package results
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/heptio/sonobuoy/pkg/plugin"
+	"github.com/heptio/sonobuoy/pkg/plugin/driver/daemonset"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// StatusFailed is the key we base junit pass/failure off of and save into
+	// our canonical results format.
+	StatusFailed = "failed"
+
+	// StatusPassed is the key we base junit pass/failure off of and save into
+	// our canonical results format.
+	StatusPassed = "passed"
+
+	// StatusSkipped is the key we base junit pass/failure off of and save into
+	// our canonical results format.
+	StatusSkipped = "skipped"
+
+	// StatusUnknown is the key we fallback to in our canonical results format
+	// if another can not be determined.
+	StatusUnknown = "unknown"
+
+	// PostProcessedResultsFile is the name of the file we create when doing
+	// postprocessing on the plugin results.
+	PostProcessedResultsFile = "sonobuoy_results.yaml"
+)
+
+// ResultFormat constants are the supported values for the resultFormat field
+// which enables post processing.
+const (
+	ResultFormatJunit = "junit"
+	ResultFormatE2E   = "e2e"
+	ResultFormatRaw   = "raw"
+)
+
+// postProcessor is a function which takes two strings: the plugin directory and the
+// filepath in question, and parse it to create an Item.
+type postProcessor func(string, string) (Item, error)
+
+// fileSelector is a type of a function which, given a filename and the FileInfo will
+// determine whether or not that file should be postprocessed. Allows matching a specific
+// file only or all files with a given suffix (for instance).
+type fileSelector func(string, os.FileInfo) bool
+
+// Item is the central format for plugin results. Various plugin
+// types can be transformed into this simple format and set at a standard
+// location in our results tarball for simplified processing by any consumer.
+type Item struct {
+	Name     string            `json:"name" yaml:"name"`
+	Status   string            `json:"status" yaml:"status"`
+	Metadata map[string]string `json:"meta,omitempty" yaml:"meta,omitempty"`
+	File     string            `json:"file,omitempty" yaml:"file,omitempty"`
```

Comment:
Is `File` needed if we're also including the path within `Metadata`? It's only being set when processing raw files. I'm not sure if it should only be included there or if it's being missed from the junit processor.

Reply:
So the `File` field is meant to be a breadcrumb for raw files, currently, an item should set either `File` or `Items` but not both.

The metadata field was added later and the idea of putting the full filepath made sense for tests since it wasn't listed elsewhere. You're right that it is now redundant for the `raw` type.

I think you're right that we can remove it, let me work on that change. Instead of checking if `.File == ""` we will just check `if .Metadata!=nil && .Metadata["file"] != ""`. It is a bit more verbose but it makes a single source of truth for the filename.